### ISSUE TITLE
Improve Documentation & Auto-Generate Documentation

### ISSUE THREAD

**Issue by Mytherin** [duckdb/duckdb#2097]
The documentation on our website is rather outdated at this point, and rather sparse particularly for the various client connectors. We should make more of an effort to improve the documentation.

One aspect in which we should be able to improve this without continuous manual labor is to auto-generate specific subsets of the documentation. Specifically, we should be able to extract documentation for the following entries in the code:

* Scalar functions
* Aggregates/Window functions
* Pragma functions
* Built-in views
* Built-in macro functions
* API Functions (C++ API, C API, Python API, R API)

For our built-in functions, we can inline our current documentation into the functions itself. For example, the documentation for `substring` is this:

|             Function             |                                                                   Description                                                                   |         Example          | Result |
|----------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|--------|
| substring(string, start, length) | Extract substring of length characters starting from character start. Note that a start value of 1 refers to the first character of the string. | substring('Hello', 2, 2) | el     |

We can then have the following struct, together with a function that returns the docs for each SQL function:

```cpp
struct FunctionDocumentation {
    string description;
    vector<string> parameters;
    string example;
};

FunctionDocumentation SubstringFun::GetDocumentation() {
    FunctionDocumentation result;
    result.description = "Extract substring of length characters starting from character start. Note that a start value of 1 refers to the first character of the string.";
    result.parameters = { "string", "start", "length" };
    result.example = "substring('Hello', 2, 2)";
    return result;
}
```

The nice thing here is that the documentation is kept alongside the actual functions, and will hence (hopefully) not go out of date as fast. We can also retrieve the documentation from inside the engine itself, e.g. we can envision allowing `DESCRIBE` on functions:

```sql
DESCRIBE substring;
┌───────────┬─────────────┬─────────────────────────────────────────────────┬──────────────────────────────────────────────────────────────────────────────────┬──────────────────────────┐
│   name    │ return_type │                   parameters                    │                                   description                                    │         example          │
├───────────┼─────────────┼─────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────┼──────────────────────────┤
│ substring │ VARCHAR     │ [string VARCHAR, start INTEGER, offset INTEGER] │ Extract substring of length characters starting from character start. Note th... │ substring('Hello', 2, 2) │
└───────────┴─────────────┴─────────────────────────────────────────────────┴──────────────────────────────────────────────────────────────────────────────────┴──────────────────────────┘
```

**Comment by Mause** [comment]
@Mytherin part of this is being done in your core function split, correct?

**Comment by Mytherin** [comment]
Yes, see #7310. What remains after that is generating the actual documentation for the website from there
